code/modules/library/library_computer.dm

```diff
@@ -0,0 +1,577 @@
+#define LIBRARY_BOOKS_PER_PAGE 25
+///Login state for our computer, this state grants full access to functions
+#define LOGIN_FULL   1
+///Login state for our computer, this state grants basic access to functions
+#define LOGIN_PUBLIC 2
+///Wait time before printing another book, used to prevent spam
+#define PRINTING_COOLDOWN (5 SECONDS)
+
+/**
+  * # Library Computer
+  *
+  * This is the player facing machine that handles all library functions
+  *
+  * This holds all procs for handling book checkins/checkout, book fines, book obj creation/modification
+  * the object also holds static lists for book inventory and checkouts. NO SQL CALLS OR QUERIES ARE MADE HERE, all
+  * of those are handled by the global library catalog that we will reference, and it should stay that way :)
+  */
+/obj/machinery/computer/library
+	name = "Library Computer"
+	anchored = TRUE
+	density = TRUE
+	icon_keyboard = null
+	icon_screen = "computer_on"
+	icon = 'icons/obj/library.dmi'
+	icon_state = "computer"
+
+	//We define a required access only to lock library specific actions like ordering/managing books to librarian access+
+	req_one_access = list(ACCESS_LIBRARY)
+	///Page Number for going through player book archives
+	var/archive_page_num = 1
+	///report category_id we have selected
+	var/selected_report
+	///Total number of pages for the parameters have set for our booklist
+	var/num_pages = 0
+	///total inventoried books, used for setting book library IDs
+	var/total_books = 0
+	///list for storing player inputs and selections, helpful for cutting down on single variable declarations
+	var/datum/library_user_data/user_data = new()
+	///This list temporarily stores the player books we grab from the DB in datums, we only update it when we need to for performance reasons
+	var/list/cached_booklist = list()
+	///Static List of borrowbook datums, used to track book checkouts acrossed the library system
+	var/static/list/checkouts = list()
+	///Static List of book datums to track what books the librarian has added to the library inventory
+	var/static/list/inventory = list()
+	///How Long a book is allowed to be checked out for
+	var/checkoutperiod = 15 MINUTES
+	///Wait period for printing books
+	var/print_cooldown = 5 SECONDS
+
+
+/obj/machinery/computer/library/Initialize(mapload)
+	. = ..()
+	populate_booklist(async = FALSE)
+	//since ui_data screws up when SQL calls are made inside it,
+	//we must populate our booklist before ui_act is called for the first time
+
+/obj/machinery/computer/library/attack_ai(mob/user)
+	return attack_hand(user)
+
+/obj/machinery/computer/library/attack_hand(mob/user)
+	if(..())
+		return
+	ui_interact(user)
+
+/obj/machinery/computer/library/attack_ghost(mob/user)
+	ui_interact(user)
+
+/obj/machinery/computer/library/attackby(obj/item/O, mob/user, params)
+	if(istype(O, /obj/item/book))
+		select_book(O)
+		return
+	if(istype(O, /obj/item/barcodescanner))
+		var/obj/item/barcodescanner/B = O
+		B.computer = src
```

Comment:
Smells like GC problems. Be sure to handle them properly. This can be done with signals and the like. Easiest is prob to make a proc for adding the PC and removing it.

Reply:
good catch

Reply:
adds proper GC'ing here